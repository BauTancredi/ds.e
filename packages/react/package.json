{
  "name": "@ds.e/react",
  "version": "1.0.0",
  "main": "lib/index.js",
  "license": "MIT",
  "dependencies": {
    "react": "^17.0.2"
  },
  "devDependencies": {
    "@babel/preset-env": "^7.16.5",
    "@babel/preset-react": "^7.16.5",
    "@babel/preset-typescript": "^7.16.5",
    "@testing-library/jest-dom": "^5.16.1",
    "@testing-library/react": "^12.1.2",
    "@types/jest": "^27.0.3",
    "@types/react": "^17.0.38",
    "jest": "^27.4.5",
    "rollup": "^2.61.1",
    "rollup-plugin-typescript2": "^0.31.1",
    "typescript": "^4.5.4"
  },
  "scripts": {
    "build": "rollup -c",
    "dev": "yarn build --watch",
    "test": "jest --verbose",
    "test:watch": "yarn test -w"
  }
}
