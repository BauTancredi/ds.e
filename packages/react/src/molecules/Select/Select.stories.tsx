import React from "react";

import Select from "./Select";

import "@ds.e/scss/lib/Select.css";

const options = [
  {
    value: "option-1",
    label: "Option 1",
  },
  {
    value: "option-2",
    label: "Option 2",
  },
  {
    value: "option-3",
    label: "Option 3",
  },
];

export default {
  title: "Molecules/Select",
};

export const Common = () => <Select options={options} />;

export const RenderOption = () => (
  <Select
    options={options}
    renderOption={({ option, getOptionRecommendedProps, isSelected }) => (
      <span {...getOptionRecommendedProps({})}>
        {option.label} {isSelected ? "SELECTED!" : ""}
      </span>
    )}
  />
);

export const CustomLabel = () => (
  <Select options={options} label="Custom label here" />
);
