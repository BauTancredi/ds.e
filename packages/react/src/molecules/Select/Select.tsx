import React, { useState, useRef, useEffect } from "react";

import Text from "../../atoms/Text";

interface SelectOption {
  label: string;
  value: string;
}

interface RenderOptionProps {
  isSelected: boolean;
  option: SelectOption;
  getOptionRecommendedProps: (overrideProps: Object) => Object;
}

interface SelectProps {
  onOptionSelected?: (option: SelectOption, optionIndex: number) => void;
  options?: SelectOption[];
  label?: string;
  renderOption?: (props: RenderOptionProps) => React.ReactNode;
}

const Select: React.FC<SelectProps> = ({
  options = [],
  label = "Please select an option ...",
  onOptionSelected: handler,
  renderOption,
}) => {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const [selectedIndex, setSelectedIndex] = useState<null | number>(null);
  const labelRef = useRef<HTMLButtonElement>(null);
  const [overlayTop, setOverlayTop] = useState<number>(0);

  const onOptionSelected = (option: SelectOption, optionIndex: number) => {
    if (handler) handler(option, optionIndex);

    setSelectedIndex(optionIndex);
    setIsOpen(false);
  };

  const onLabelClicked = () => {
    setIsOpen(!isOpen);
  };

  useEffect(() => {
    setOverlayTop((labelRef.current?.offsetHeight || 0) + 10);
  }, [labelRef.current?.offsetHeight]);

  let selectedOption = null;

  if (selectedIndex !== null) selectedOption = options[selectedIndex];
  return (
    <div className="dse-select">
      <button
        ref={labelRef}
        className="dse-select__label"
        onClick={() => onLabelClicked()}
      >
        <Text>{selectedIndex === null ? label : selectedOption?.label}</Text>
        <svg
          className={`dse-select__caret ${
            isOpen ? "dse-select__caret--open" : "dse-select__caret--closed"
          }`}
          width="1rem"
          height="1rem"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            d="M19 9l-7 7-7-7"
          />
        </svg>
      </button>
      {isOpen ? (
        <ul style={{ top: overlayTop }} className="dse-select__overlay">
          {options.map((option, optionIndex) => {
            const isSelected = selectedIndex === optionIndex;

            const renderOptionProps = {
              option,
              isSelected,
              getOptionRecommendedProps: (overrideProps = {}) => {
                return {
                  className: `dse-select__option ${
                    isSelected ? "dse-select__option--selected" : ""
                  }`,
                  key: option.value,
                  onClick: () => onOptionSelected(option, optionIndex),
                  ...overrideProps,
                };
              },
            };

            if (renderOption) {
              return renderOption(renderOptionProps);
            }

            return (
              <li
                onClick={() => onOptionSelected(option, optionIndex)}
                key={option.value}
                className={`dse-select__option ${
                  isSelected ? "dse-select__option--selected" : ""
                }`}
              >
                <Text>{option.label}</Text>
                {isSelected ? (
                  <svg
                    width="1rem"
                    height="1rem"
                    className="w-6 h-6"
                    fill="none"
                    stroke="currentColor"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth={2}
                      d="M5 13l4 4L19 7"
                    />
                  </svg>
                ) : null}
              </li>
            );
          })}
        </ul>
      ) : null}
    </div>
  );
};

export default Select;
