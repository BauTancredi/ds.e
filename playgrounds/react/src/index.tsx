import React from "react";
import ReactDOM from "react-dom";

import { Select } from "@ds.e/react";

import "@ds.e/scss/lib/Utilites.css";
import "@ds.e/scss/lib/Text.css";
import "@ds.e/scss/lib/Margin.css";
import "@ds.e/scss/lib/Select.css";
import "@ds.e/scss/lib/global.css";

const options = [
  {
    value: "option-1",
    label: "Option 1",
  },
  {
    value: "option-2",
    label: "Option 2",
  },
  {
    value: "option-3",
    label: "Option 3",
  },
];

ReactDOM.render(
  <div style={{ padding: "40px" }}>
    <Select options={options} />
  </div>,
  document.querySelector("#root")
);
