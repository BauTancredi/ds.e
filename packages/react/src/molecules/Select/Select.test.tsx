import React from "react";
import { render, fireEvent } from "@testing-library/react";

import Select from "./Select";

const options = [
  {
    value: "option-1",
    label: "Option 1",
  },
  {
    value: "option-2",
    label: "Option 2",
  },
  {
    value: "option-3",
    label: "Option 3",
  },
];

describe("Select", () => {
  it("renders all options passed to it", () => {
    const { getAllByRole, getByTestId } = render(<Select options={options} />);

    fireEvent.click(getByTestId("DseSelectButton"));

    expect(getAllByRole("menuitemradio")).toHaveLength(options.length);
  });

  it("renders options using custom renderOption method if passed as prop", () => {
    const { getAllByTestId, getByTestId } = render(
      <Select
        options={options}
        renderOption={({ option, getOptionRecommendedProps }) => {
          return (
            <li
              data-testid="CustomRenderOption"
              {...getOptionRecommendedProps({})}
            >
              {option.label}
            </li>
          );
        }}
      />
    );

    fireEvent.click(getByTestId("DseSelectButton"));

    expect(getAllByTestId("CustomRenderOption")).toHaveLength(options.length);
  });

  it("calls the onOptionSelected prop with the selected options and its index if passed", () => {
    const onOptionSelected = jest.fn();
    const { getAllByRole, getByTestId } = render(
      <Select options={options} onOptionSelected={onOptionSelected} />
    );

    fireEvent.click(getByTestId("DseSelectButton"));

    fireEvent.click(getAllByRole("menuitemradio")[0]);

    expect(onOptionSelected).toHaveBeenCalledWith(options[0], 0);
  });

  it("the button label changes to the selected option label", () => {
    const { getAllByRole, getByTestId } = render(<Select options={options} />);

    fireEvent.click(getByTestId("DseSelectButton"));

    fireEvent.click(getAllByRole("menuitemradio")[0]);

    expect(getByTestId("DseSelectButton")).toHaveTextContent(options[0].label);
  });

  it("can customize select label", () => {
    const { getByText } = render(
      <Select options={options} label="This is a custom label" />
    );

    expect(getByText(/This is a custom label/)).toBeInTheDocument();
  });

  it("snapshot of the selected option state", () => {
    const { getAllByRole, getByTestId, asFragment } = render(
      <Select options={options} />
    );

    fireEvent.click(getByTestId("DseSelectButton"));

    fireEvent.click(getAllByRole("menuitemradio")[0]);

    expect(asFragment()).toMatchSnapshot();
  });

  it("snapshot of the base state", () => {
    const { asFragment } = render(<Select options={options} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("snapshot of the options menu open states", () => {
    const { getByTestId, asFragment } = render(<Select options={options} />);

    fireEvent.click(getByTestId("DseSelectButton"));

    expect(asFragment()).toMatchSnapshot();
  });
});
