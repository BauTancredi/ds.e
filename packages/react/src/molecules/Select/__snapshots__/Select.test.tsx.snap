// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select snapshot of the base state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="ds-select-list"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-text-base"
      >
        Please select an option ...
      </p>
      <svg
        class="dse-select__caret dse-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Select snapshot of the options menu open states 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="ds-select-list"
      aria-expanded="true"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-text-base"
      >
        Please select an option ...
      </p>
      <svg
        class="dse-select__caret dse-select__caret--open"
        fill="none"
        height="1rem"
        stroke="currentColor"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
    <ul
      class="dse-select__overlay"
      id="ds-select-list"
      role="menu"
      style="top: 10px;"
    >
      <li
        aria-label="Option 1"
        class="dse-select__option 
                  
                  
                  }"
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Option 1
        </p>
      </li>
      <li
        aria-label="Option 2"
        class="dse-select__option 
                  
                  
                  }"
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Option 2
        </p>
      </li>
      <li
        aria-label="Option 3"
        class="dse-select__option 
                  
                  
                  }"
        role="menuitemradio"
        tabindex="0"
      >
        <p
          class="dse-text dse-text-base"
        >
          Option 3
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Select snapshot of the selected option state 1`] = `
<DocumentFragment>
  <div
    class="dse-select"
  >
    <button
      aria-controls="ds-select-list"
      aria-haspopup="true"
      class="dse-select__label"
      data-testid="DseSelectButton"
    >
      <p
        class="dse-text dse-text-base"
      >
        Option 1
      </p>
      <svg
        class="dse-select__caret dse-select__caret--closed"
        fill="none"
        height="1rem"
        stroke="currentColor"
        viewBox="0 0 24 24"
        width="1rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
